starfish/codebook.py

```diff
@@ -372,51 +426,26 @@ def decode_euclidean(self, intensities: IntensityTable) -> IntensityTable:
 
         """
 
-        def _min_euclidean_distance(observation: xr.DataArray, codes: Codebook) -> np.ndarray:
-            """find the code with the closest euclidean distance to observation
-
-            Parameters
-            ----------
-            observation : xr.DataArray
-                2-dimensional DataArray of shape (n_ch, n_hyb)
-            codes :
-                Codebook containing codes to compare to observation
+        intensity_mask = intensities.mask_low_intensity_features(min_intensity)
 
-            Returns
-            -------
-            np.ndarray :
-                1-d vector containing the distance of each code to observation
+        # normalize both the intensities and the codebook
```

Comment:
from an offline convo -- i suspect this is one potential source of your MERFISH results not more closely matching the MERFISH results on master. in the MERFISH pipeline, codes and intensities are re-scaled by their L2 norms, not by their sums. In other words, these re-scaled values have unit _magnitudes_ not unit _sums_. The old code went something like this:

```    @staticmethod
    def _normalize_barcodes(codebook: pd.DataFrame) -> np.ndarray:
        # parse barcode into numpy array and normalize by l2_norm
        codes = np.array([np.array([int(d) for d in c]) for c in codebook.barcode])
        codes_l2_norm = np.linalg.norm(codes, axis=1, ord=2)
        weighted_codes = codes / codes_l2_norm[:, None]
        return weighted_codes

    # TODO ambrosejcarr fill in these values + add docs
    @staticmethod
    def _parse_pixel_traces(encoded: pd.DataFrame) -> Tuple[np.ndarray, np.ndarray]:
        # parse spots into pixel traces, normalize and filter
        df = encoded.loc[:, ['spot_id', 'barcode_index', 'intensity']]
        # TODO this assumes that bits are sorted [they are, currently]
        pixel_traces = df.pivot(index='spot_id', columns='barcode_index', values='intensity')
        pixel_traces = pixel_traces.values
        pixel_traces_l2_norm = np.linalg.norm(pixel_traces, axis=1, ord=2)
        ind = pixel_traces_l2_norm > 0
        pixel_traces[ind, :] = pixel_traces[ind, :] / pixel_traces_l2_norm[ind, None]
        return pixel_traces, pixel_traces_l2_norm
```

Reply:
in fact, you can generalize the above code to do 'unit_normalize' by changing the 'ord' parameter in np.linalg.norm to 1. hence the generalization is incurred by writing a method called norm, which exposes a parameter ord, which is then imply passed to np.linalg.norm.

Reply:
Nice, thanks for the comments. I'll follow up on this and see if it decreases the discrepancy (my plan was to attack that in person this week)